### PR TITLE
Ensure the line feed follows carriage return when reading from the DOS console

### LINKED ISSUE #4007: Handling of \r and \n in DOS function 3Fh
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

dosbox-staging, version 0.83.0-alpha (38b98)

### What Operating System are you using?

Windows 11

### If Other OS, please describe

_No response_

### Relevant hardware info

_No response_

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

When trying to play the "game" [Ford Fiesta Werbespiel FIESTA.ZIP](https://github.com/user-attachments/files/17528178/Ford.Fiesta.Werbespiel.FIESTA.ZIP), you cannot get past the name entry.

An analysis by someone who tried it was the following:
>  The game reads keyboard input using DOS Function 3Fh, 1 byte at a time, from handle #1. It expects to first read 0Dh, then 0Ah. DOSBox will return 0D 0A when reading two bytes at once from STDIN. But when reading one byte at a time, DOSBox' emulated DOS will only return the 0D and drop the 0A.

and

> The game works without issue when first booting real PC-DOS, then switching over to the demo disk image and running it in real PC-DOS, in DOSBox as well as 86Box.

### Steps to reproduce the behaviour.

1. Start the game
2. press "space" until it asks you to enter your name
3. enter anything and press [return]
4. note that the cursor just goes to the next line and you can continue typing input, the game does not proceed

### Download URL of affected game or software

_No response_

### Your configuration

_No response_

### Provide a Log

```text
(no relevant entries in the log)
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

### PR DESCRIPTION
**Pull request by interloper98**
# Description

When reading the MS-DOS console device, the line feed character always follows a carriage return (the infamous Microsoft 'CRLF' sequence :grin: ).

DOSBox gets this right when reading enough bytes at a time that both the CRLF can fit into a single read operation. However, if the read size is split or done byte-by-byte, the trailing LF isn't sent.

Thanks to @darkstar's issue report and their acquaintance's description of the cause.

This must not have been a big deal given it hasn't posed a problem in decades, but a German adware game by Ford Motor Company is sensitive to this and always expects CRLF after pressing enter.

Reviewers - I had to apply clang-format to cleanup `dev_con.h` first, so that's the first commit. Suggest skipping it and just reading the second and third commits.

## Related issues

Fixes #4007

# Release notes

CRLF is now always sent when reading from the console device. This fixes the name entry part of the 'Ford Fiesta Werbespiel' game.

# Manual testing

TODO: Need to test more apps, games, etc. Will update as I test more.

# Checklist

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [x] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [x] provided the release notes draft (for significant user-facing changes).
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: src/dos/dev_con.h [hunk 1]
```diff
@@ -75,6 +75,14 @@ void device_CON::ClearAnsi()
 	ansi.numberofarg = 0;
 }
 
+enum ControlCode : uint8_t {
```
Comment (FeralChild64) [inline src/dos/dev_con.h:78]: Would it be feasible to move this enum somewhere to `/include/*.h` and make `program_more_output.cpp` reuse it?

Comment (interloper98) [reply src/dos/dev_con.h:78]: Oh, nice - I didn't see that block of constants. Thanks for this heads up, @FeralChild64.

```
// ASCII control characters
constexpr char code_ctrl_c = 0x03; // end of text
constexpr char code_bs     = 0x08; // backspace
constexpr char code_lf     = 0x0a; // line feed
constexpr char code_ff     = 0x0c; // form feed
constexpr char code_cr     = 0x0d; // carriage return
constexpr char code_esc    = 0x1b;
constexpr char code_del    = 0x7f;
```

I guess it should go in an `ascii.h` file, but for now we don't need all ~200+ characters defined - we can start with this set of control codes, and add as needed?

Will get this added and re-use it in `program_more_output.cpp`

**Comment by darkstar** [comment]
Thanks so much for working on this!

I didn't expect the fix to be such a big deal, having to restructure so much of the console code. Thanks for working on this

> Obviously this must not have been a big deal given it hasn't posed a problem in decades

I'm not sure. Not everyone has the resources to analyze and debug the core issue when some game or program they're using doesn't work as expected.... But I agree it's not really a high-profile bug :)

File: src/dos/dev_con.h (hunk 1 above)
Comment (FeralChild64) [reply src/dos/dev_con.h:78]: Definitely - no need to have all the characters for now.

**Comment by interloper98** [comment]
The description of the problem in your ticket was really the key and made it easy. I just searched where the console input was expanding the CRLF :-) Thanks for that!

Fortunately it's just a one line fix. All the other adjustments are just applying code formatting rules; and that's all automated :)

**Comment by darkstar** [comment]
> The Windows build workflow is broken right now. I guess you'll have to wait to try it. But if you have a build environment, it's in the `il98/stdin-sends-crlf-sequence` branch.

Probably for the same reason my local build is not working anymore (it also stops at glib). Something to do with Python, they removed a module (distutils) in 3.12 or 3.13 which the glib build needs

```
Traceback (most recent call last):
  File "D:\Devel\vcpkg\buildtrees\glib\x64-windows-dbg\gio\gdbus-2.0\codegen\gdbus-codegen", line 53, in <module>
    from codegen import codegen_main
  File "D:\Devel\vcpkg\buildtrees\glib\x64-windows-dbg\gio\gdbus-2.0\codegen\codegen_main.py", line 29, in <module>

    from . import dbustypes

  File "D:\Devel\vcpkg\buildtrees\glib\x64-windows-dbg\gio\gdbus-2.0\codegen\dbustypes.py", line 22, in <module>

    from . import utils

  File "D:\Devel\vcpkg\buildtrees\glib\x64-windows-dbg\gio\gdbus-2.0\codegen\utils.py", line 22, in <module>

    import distutils.version

ModuleNotFoundError: No module named 'distutils'
```

File: src/dos/dev_con.h [hunk 2]
```diff
@@ -75,6 +75,14 @@ void device_CON::ClearAnsi()
...
+	Null           = 0x0,
+	Backspace      = 0x8,
+	LineFeed       = 0xa,
+	CarriageReturn = 0xd,
```
Comment (johnnovak) [inline src/dos/dev_con.h:82]: Minor: maybe use two digits consistently? so 0x08, 0x0a, etc.

**Review by johnnovak — APPROVED** [review]
Looks good to me @interloper98. Most games use the I/O ports to communicate with the keyboard controller directly, so this is relatively low risk. Plus yeah, this is the more accurate behaviour.

Maybe it's time to start an "input handling tests" page, similar to the audio and video tests I put together? E.g.,

https://github.com/dosbox-staging/dosbox-staging/wiki/Video-emulation-tests

We should put the games with the keyboard regressions after @FeralChild64's grand rework there are well that we had to fix over the past year.

**Comment by interloper98** [comment]
> I wanted to test if this solves #3173, but there is no Windows artifact.

Should be ready here! https://github.com/dosbox-staging/dosbox-staging/actions/runs/11536969579

**Comment by interloper98** [comment]
> Tested it, doesn't solve #3173.

Thanks for checking. It makes sense- that issue relates to how dosbox's own CLI shell is parsing and formatting the output and prompt. 

It's somewhat opinionated, I guess, and isn't adding an additional new line.

(Where as this PR is more about always completing the LF portion of the CRLF sequence when reading from the stdin console device).

File: src/dos/dev_con.h (hunk 2 above)
Comment (interloper98) [reply src/dos/dev_con.h]: fixed

File: src/dos/dev_con.h (hunk 1 above)
Comment (interloper98) [reply src/dos/dev_con.h]: Fixed; dropped them into an enum class `Ascii`, so we can use `Ascii::LineFeed`, etc. I tried 'ControlCode::LineFeed`, but `Ascii::` was a lot more intuitive/obvious. So went with that.

**Comment by interloper98** [comment]
> Maybe it's time to start an "input handling tests" page, similar to the audio and video tests I put together? E.g.,
> 
> https://github.com/dosbox-staging/dosbox-staging/wiki/Video-emulation-tests
> 
> We should put the games with the keyboard regressions after @FeralChild64's grand rework there are well that we had to fix over the past year.

:+1: ; added (under the `Dev` list on the right)

**Comment by johnnovak** [comment]
> > Maybe it's time to start an "input handling tests" page, similar to the audio and video tests I put together? E.g.,
> > https://github.com/dosbox-staging/dosbox-staging/wiki/Video-emulation-tests
> > We should put the games with the keyboard regressions after @FeralChild64's grand rework there are well that we had to fix over the past year.
> 
> 👍 ; added (under the `Dev` list on the right)

Great stuff, thanks man 💯

**Comment by interloper98** [comment]
Tested under various games that accept keyboard and 'Enter' for input, text mode applications, Windows 3.1x, and applications inside Windows 3.1x.  Zero problems (like double-enter or not ack'ing the enter; merging!

**Pull request merged**